packages/runtime-dom/src/directives/vOn.ts

```diff
@@ -0,0 +1,65 @@
+import { isArray } from '@vue/shared'
+
+const modifierGuards: Record<string, (e: Event) => void | boolean> = {
+  stop: e => e.stopPropagation(),
+  prevent: e => e.preventDefault(),
+  self: e => e.target !== e.currentTarget,
+  ctrl: e => !(e as any).ctrlKey,
+  shift: e => !(e as any).shiftKey,
+  alt: e => !(e as any).altKey,
+  meta: e => !(e as any).metaKey,
+  left: e => 'button' in e && (e as any).button !== 0,
+  middle: e => 'button' in e && (e as any).button !== 1,
+  right: e => 'button' in e && (e as any).button !== 2
+}
+
+const keyNames: Record<string, string | string[]> = {
+  esc: 'escape',
+  // IE11 uses `Spacebar` for Space key name.
+  space: [' ', 'spacebar'],
+  up: 'arrowup',
+  left: 'arrowleft',
+  right: 'arrowright',
+  down: 'arrowdown',
+  // IE11 uses `Del` for Delete key name.
+  delete: ['backspace', 'del']
+}
+
+const modifierKeys = ['ctrl', 'shift', 'alt', 'meta']
+
+export const vOnModifiersGuard = (fn: Function, modifiers: string[]) => {
+  return (event: Event) => {
+    for (let modifier of modifiers) {
+      if (
+        modifier === 'exact' &&
+        // Some of the modifierKeys are not specified as modifier, but is flagged true on the event
+        modifierKeys.some(
+          modifierKey =>
+            modifiers.indexOf(modifierKey) === -1 &&
+            (event as any)[`${modifierKey}Key`]
+        )
+      ) {
+        return
+      }
+      const guard = modifierGuards[modifier]
+      if (guard && guard(event)) return
+    }
+    if ('key' in event) {
```

Comment:
The whole `key` logic can be extracted to a separate exported function (e.g. `guardKey`). The compiler can then only inject/wrap the handler if the event is one of `keyup`, `keydown` or `keypress`. This allows the key related logic to be tree-shaken if no key modifiers are used.